transport_plugins/jlink/iotile_transport_jlink/jlink.py

```diff
@@ -331,48 +386,41 @@ def send_rpc_async(self, conn_id, address, rpc_id, payload, timeout, callback):
             rpc_id (int): the 16-bit id of the RPC we want to call
             payload (bytearray): the payload of the command
             timeout (float): the number of seconds to wait for the RPC to execute
-            callback (callable): A callback for when we have finished the RPC.  The callback will be called as"
-                callback(connection_id, adapter_id, success, failure_reason, status, payload)
-                'connection_id': the connection id
-                'adapter_id': this adapter's id
-                'success': a bool indicating whether we received a response to our attempted RPC
-                'failure_reason': a string with the reason for the failure if success == False
-                'status': the one byte status code returned for the RPC if success == True else None
-                'payload': a bytearray with the payload returned by RPC if success == True else None
         """
 
-        def _on_finished(_name, retval, exception):
-            if exception is not None:
-                callback(conn_id, self.id, False, str(exception), None, None)
-                return
-
-            callback(conn_id, self.id, True, None, retval['status'], retval['payload'])
-
         # Default to polling for the response every 1 millisecond
         # FIXME, add an exponential polling backoff so that we wait 1, 2, 4, 8, etc ms
-        self._control_thread.command(JLinkControlThread.SEND_RPC, _on_finished, self._device_info, self._control_info, address, rpc_id, payload, 0.001, timeout)
+        self._ensure_connection(conn_id, True)
+
+        response = None
+
+        for attempt in range(retries):
+            try:
+                response = await self._jlink_async.send_rpc(
+                    self._device_info, self._control_info,
+                    address, rpc_id, payload, timeout)
+            except HardwareError as exc:
+                if exc.params['external_gate_error'] == 1:
```

Comment:
This needs to be `exc.params.get('external_gate_error')` or it will raise KeyError when external_gate_error is not passed (so for any other HardwareError).  `get` from a dict will safely return `None` if the key is not found rather than raising.

Reply:
Fixed.